std/algorithm/comparison.d

```diff
@@ -892,116 +892,367 @@ nothrow pure @safe unittest
         }
-        // if one of the arguments is a string and the other isn't, then auto-decoding
-        // can be avoided if they have the same ElementEncodingType
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isAutodecodableString!Range1 != isAutodecodableString!Range2 &&
-            is(ElementEncodingType!Range1 == ElementEncodingType!Range2))
+        else
         {
-            import std.utf : byCodeUnit;
+            import std.meta : staticMap, staticIndexOf;
+
+            // get lengths
+            alias hasLengthBits = staticMap!(hasLength, Rs);
+            enum indexOfFirstLength = staticIndexOf!(true, hasLengthBits);
+
+            // if `true`, at least one of `Rs` has a `length` property
+            enum someHasLength = indexOfFirstLength != -1;
 
-            static if (isAutodecodableString!Range1)
+            static if (someHasLength && // if one range has a (finite) length and
+                       anySatisfy!(isInfinite, Rs)) // another is infinite
             {
-                return equal(r1.byCodeUnit, r2);
+                return false;   // they can't be equal
             }
             else
             {
-                return equal(r2.byCodeUnit, r1);
-            }
-        }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
-        {
-            immutable len1 = r1.length;
-            immutable len2 = r2.length;
-            if (len1 != len2) return false; //Short circuit return
+                enum someHasIndexingAndLength = anySatisfy!(hasIndexingAndLength, Rs);
 
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
-            {
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
-            }
-            return true;
-        }
-        else
-        {
-            //Generic case, we have to walk both ranges making sure neither is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
-            {
-                if (r2.empty) return false;
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                static if (someHasLength)
+                {
+                    /* if any `rs` `hasLength` then `primaryRangeIndex` will index the
+                     * first of them */
+                    enum primaryRangeIndex = indexOfFirstLength;
+                    const length = rs[indexOfFirstLength].length;
+                }
+                else
+                {
+                    enum primaryRangeIndex = 0; // just pick first
+                }
+                alias primaryRange = rs[primaryRangeIndex]; // shorthand
+
+                // check lengths
```

Comment:
Maybe explain expand the comment a bit more explaining that this is an optional optimization for skipping the comparison before it even started.